### ISSUE TITLE
Jet Set Radio Future [JSRF]: Incorrect rendering of progress indicator panel

### ISSUE THREAD

**Issue by Shoegzer** [xemu-project/xemu#601]
### Bug Description

The progress indicator in the lower-left panel of the map screen in JSRF is black (it should be light-blue) and icons are not visible.  Borrowing screenshots because laziness:

Incorrect:
![124822410-db4a2280-df3d-11eb-9566-f58de8b48250](https://user-images.githubusercontent.com/12475371/146827688-9b0e1734-889b-48b4-b288-b8a3ca462c09.png)
Correct:
![146447186-d617a735-a2f3-485c-9f6f-3d21f95f26f8](https://user-images.githubusercontent.com/12475371/146827702-20d7824f-137d-44b1-9876-a4f61f906d51.png)

Note: The issue with the map itself being incorrectly rendered as solid grey is being tracked in #599.

### Expected Behavior

Panel should be light-blue and icons should be visible.

### xemu Version

0.6.2-10-gaa7acc8f64

### System Information

OS: Linux Mint 20.2 / kernel 5.13
CPU: Intel i7-10870H
GPU: Nvidia GTX 3070
GPU Driver: 470.86

**Comment by GXTX** [comment]
Dupe of #397

**Comment by Shoegzer** [comment]
It isn't though.  See note above, the map issue is being tracked separately and @abaire requested a separate issue report for the info panel.

**Comment by abaire** [comment]
Tracking some initial findings:

- I see some unhandled pgraph messages that I think are related to specular lighting, but I wouldn't expect that to affect this. To more or less rule it out, I modified the shader to set `oD0`'s red channel to 0.8 and the resultant image is still black.
- The icons are missing/black as well as the blue background so, coupled with the lack of effect when hardcoding the vertex colors, I suspect this may be a texturing issue. I'm guessing the info panel may be a textured quad with dynamic text rendered on top of it, and that there may be some issue with the fixed function shader that is preventing the correct texture from being utilized (perhaps due to some state leakage since textures clearly work elsewhere on this screen?). Also possible that the texture is never loaded properly in the first place or is in some unusual format that is mishandled, but given that other textures on the screen work it seems unlikely that the developers would've used a special format just for this element.

**Comment by abaire** [comment]
Looking some more, I see groups of 3 `ARRAY_ELEMENT16` calls (= 6 vertex indices = 1 quad) that are preceded by calls that disable lighting altogether and set up the FVF in position + texcoord mode so I'm relatively confident that this problem has nothing to do with lighting/materials.

**Comment by abaire** [comment]
And some more poking around: I can see a second block of two quads being rendered that have normals (the first set consists of 11 quads that do not have normals).

With #603 I can see that the textures that work show up correctly
![Screenshot_20211224_081130](https://user-images.githubusercontent.com/448413/147364108-c0b8ae7f-f47d-41b1-8ea4-a671890a1390.png)

but I don't see the texture for the panel that's missing (there are a couple draw calls that appear to not be enabling textures in xemu at all). So I suspect there's some issue where xemu isn't doing the glBind necessary to get the texture to display.

**Comment by abaire** [comment]
Maybe I take it back. The only place on that screen that looks like it'd use the rounded gradient under the SS/S/M/L/XL box is this panel, and maybe the red dot is just a recoloring of the blue dot.

**Comment by abaire** [comment]
Spent a little more time on this. I can see that the panel is rendered with the correct texture selected just after doing the upper right hand panel with the trophy icon and the "TOTAL" and "x" static text.

It looks like the panel is a 54 element `TRIANGLE` draw command and unless there's a bug in the debugger (entirely possible at this point), the correct texture is still displayed when it is rendered.

I'm reasonably confident that the broken panel is drawn with this set of pgraph commands:
[jsrf_black_panel.txt](https://github.com/mborgerson/xemu/files/7777704/jsrf_black_panel.txt)

**Comment by abaire** [comment]
It looks like lighting is enabled just before the panel is drawn, and specular lighting is enabled with light 0 in LIGHT_INFINITE, which is at least not fully implemented (there's a `/* FIXME: Do specular */`).

In comparison, the upper-right panel has lighting disabled completely, which probably explains why it works.

This is a dump with both the upper right and lower left panels.
[jsrf_black_panel.txt](https://github.com/mborgerson/xemu/files/7777796/jsrf_black_panel.txt)

**Comment by abaire** [comment]
These are the shaders in use
[broken_call.txt](https://github.com/mborgerson/xemu/files/7778012/broken_call.txt)

**Comment by abaire** [comment]
Narrowed it down, there's an issue with the normal calculation that's multiplied by the light/material diffuse color.

If I force `nDotVP` to 1.0, the panel displays more or less as expected (I assume it may be too bright if the normal is not actually pointed straight at the light).

https://github.com/mborgerson/xemu/blob/master/hw/xbox/nv2a/shaders.c#L556

Will need to come back to this when I have more time and figure out what exactly is going wrong.

**Comment by abaire** [comment]
Little more digging, it looks like the BEGIN_END that actually renders the panel is not passing a normal along (I see it sets the offset for the position and texture coords but nothing else). I think that means that when it gets into the shader the normal is 0, so the light will never work. I'll build a test for the HW to see if there's some magic default normal or if it reuses something.

**Comment by abaire** [comment]
Very surprisingly, it looks like the normals from the previous draw call are preserved on HW, whereas there are not with xemu.

JSRF makes a very interesting draw just before starting to render the bottom left panel. Lighting is disabled, but the call passes normals for a quad (2 TRIANGLE elements). It then enables lighting and draws a relatively complicated mesh for the broken panel.

I wrote a quick test that does something similar, rendering an untextured, unlit triangle that passes normals, then rendering an untextured, lit triangle that passes no normals. On HW, this results in a lit second triangle. On xemu this results in a black second triangle.

I think this is the culprit, will just have to figure out the right way to reproduce this behavior and confirm whether it's just keeping the last set normal or doing something more complicated.

**Comment by abaire** [comment]
Looks pretty convincing that it just reuses whatever was set as the last normal, so this should be straightforward to implement in xemu.

**Comment by abaire** [comment]
Test is here: https://github.com/abaire/nxdk_pgraph_tests/blob/main/tests/lighting_normal_tests.cpp and confirms that the value of the last set normal is applied when calculating lighting for vertices that do not set a normal.

**Comment by abaire** [comment]
Hey @mborgerson - I think fixing this would involve saving the last normal on any `BEGIN_END` w/ normals in the buffer (and then piping that through the shader). I believe this will be straightforward but a bit messy since there are a number of different buffer source options; do you have any suggestions for a better mechanism?
